File: android/src/main/java/expo/modules/footshopreactnativesharedstorage/SharedStorageModule.kt
package expo.modules.footshopreactnativesharedstorage

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.modules.Promise

data class SetOptions(
    @Field
    var storageKey: String = "",

    @Field
    var key: String = "",

    @Field
    var data: Any? = null
)

data class GetOptions(
    @Field
    var storageKey: String = "",

    @Field
    var key: String = ""
)

data class GetAllKeysOptions(
    @Field
    var storageKey: String = ""
)

data class ContainsOptions(
    @Field
    var storageKey: String = "",

    @Field
    var key: String = ""
)

class SharedStorageModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("SharedStorage")

        AsyncFunction("get") { options: GetOptions, promise: Promise ->
            launch(Dispatchers.Main) {
                val sharedPreferences = getSharedPreferences(options.storageKey)
                if (options.key.isNotBlank()) {
                    try {
                        val value = sharedPreferences.getString(options.key, null)
                        promise.resolve(value)
                    } catch (e: Exception) {
                        promise.reject("GET_ERROR", "An error occurred while retrieving data.", e)
                    }
                } else {
                    promise.reject("INVALID_OPTIONS", "The 'key' option is missing.")
                }
            }
        }

        AsyncFunction("set") { options: SetOptions, promise: Promise ->
            launch(Dispatchers.Main) {
                val sharedPreferences = getSharedPreferences(options.storageKey)
                if (options.key.isNotBlank() && options.data != null) {
                    try {
                        val editor = sharedPreferences.edit()
                        editor.putString(options.key, options.data.toString())
                        editor.apply()
                        promise.resolve(null)
                    } catch (e: Exception) {
                        promise.reject("SET_ERROR", "An error occurred while saving data.", e)
                    }
                } else {
                    promise.reject("INVALID_OPTIONS", "The 'key' and 'data' options are required.")
                }
            }
        }

        AsyncFunction("remove") { options: RemoveOptions, promise: Promise ->
            launch(Dispatchers.Main) {
                val sharedPreferences = getSharedPreferences(options.storageKey)
                if (options.key.isNotBlank()) {
                    try {
                        val editor = sharedPreferences.edit()
                        editor.remove(options.key)
                        editor.apply()
                        promise.resolve(null)
                    } catch (e: Exception) {
                        promise.reject("REMOVE_ERROR", "An error occurred while removing data.", e)
                    }
                } else {
                    promise.reject("INVALID_OPTIONS", "The 'key' option is missing.")
                }
            }
        }

        AsyncFunction("getAllKeys") { options: GetAllKeysOptions, promise: Promise ->
            launch(Dispatchers.Main) {
                val sharedPreferences = getSharedPreferences(options.storageKey)
                try {
                    val allKeys = sharedPreferences.all.keys.toList()
                    promise.resolve(allKeys)
                } catch (e: Exception) {
                    promise.reject("GET_ALL_KEYS_ERROR", "An error occurred while retrieving keys.", e)
                }
            }
        }

        AsyncFunction("contains") { options: ContainsOptions, promise: Promise ->
            launch(Dispatchers.Main) {
               val sharedPreferences = getSharedPreferences(options.storageKey)
               if (options.key.isNotBlank()) {
                   try {
                       val containsKey = sharedPreferences.contains(options.key)
                       promise.resolve(containsKey)
                   } catch (e: Exception) {
                       promise.reject("CONTAINS_ERROR", "An error occurred while checking for key existence.", e)
                   }
               } else {
                   promise.reject("INVALID_OPTIONS", "The 'key' option is missing.")
               }
           }
        }
    }

    private fun getSharedPreferences(storageKey: String): SharedPreferences {
        val context = getContext()
        return context.getSharedPreferences(storageKey, Context.MODE_PRIVATE)
    }
}

File: ios/SharedStorageModule.swift
import ExpoModulesCore

struct SetOptions: Record {
    @Field
    var storageKey: String = ""

    @Field
    var key: String = ""

    @Field
    var data: Any?
}

struct GetOptions: Record {
    @Field
    var storageKey: String = ""

    @Field
    var key: String = ""
}

struct RemoveOptions: Record {
    @Field
    var storageKey: String = ""

    @Field
    var key: String = ""
}

struct GetAllKeysOptions: Record {
    @Field
    var storageKey: String = ""
}

struct ContainsOptions: Record {
    @Field
    var storageKey: String = ""

    @Field
    var key: String = ""
}

public class SharedStorageModule: Module {
    public func definition() -> ModuleDefinition {
        Name("SharedStorage")

        AsyncFunction("set") { (options: SetOptions, promise: Promise) -> Void in
            guard let sharedDefaults = UserDefaults(suiteName: options.storageKey) else {
                promise.reject("SET_ITEM_ERROR", "User defaults is nil")
                return
            }

            sharedDefaults.setValue(options.data, forKey: options.key)
            sharedDefaults.synchronize()

            switch options.data {
            case let jsonString as String:
                promise.resolve(jsonString)
            case let jsonDictionary as [String: Any]:
                if let jsonData = try? JSONSerialization.data(withJSONObject: jsonDictionary),
                   let jsonString = String(data: jsonData, encoding: .utf8) {
                    promise.resolve(jsonString)
                } else {
                    promise.reject("SET_ITEM_ERROR", "Failed to store data")
                }
            default:
                promise.reject("SET_ITEM_ERROR", "Failed to store data")
            }
        }.runOnQueue(.main)

        AsyncFunction("get") { (options: GetOptions, promise: Promise) -> Void in
            guard let sharedDefaults = UserDefaults(suiteName: options.storageKey) else {
                promise.reject("GET_ITEM_ERROR", "User defaults is nil")
                return
            }

            guard let storedData = sharedDefaults.value(forKey: options.key) else {
                promise.reject("GET_ITEM_ERROR", "Data not found")
                return
            }

            if let jsonString = storedData as? String {
                promise.resolve(jsonString)
            } else if let jsonData = storedData as? Data, let jsonString = String(data: jsonData, encoding: .utf8) {
                promise.resolve(jsonString)
            } else {
                promise.reject("GET_ITEM_ERROR", "Failed to retrieve data")
            }
        }.runOnQueue(.main)

        AsyncFunction("remove") { (options: RemoveOptions, promise: Promise) -> Void in
            if let sharedDefaults = UserDefaults.init(suiteName: options.storageKey) {
                sharedDefaults.removeObject(forKey: options.key)
                sharedDefaults.synchronize()
                promise.resolve(nil)
            } else {
                promise.reject("DELETE_ITEM_ERROR", "User defaults is nil")
            }
        }.runOnQueue(.main)

        AsyncFunction("getAllKeys") { (options: GetAllKeysOptions, promise: Promise) -> Void in
            guard let sharedDefaults = UserDefaults(suiteName: options.storageKey) else {
                promise.reject("GET_ALL_KEYS_ERROR", "User defaults is nil")
                return
            }

            let allKeys = sharedDefaults.dictionaryRepresentation().keys.map { $0 as String }
            promise.resolve(allKeys)
        }.runOnQueue(.main)

        AsyncFunction("contains") { (options: ContainsOptions, promise: Promise) -> Void in
            guard let sharedDefaults = UserDefaults(suiteName: options.storageKey) else {
                promise.reject("CONTAINS_ERROR", "User defaults is nil")
                return
            }

            let containsKey = sharedDefaults.object(forKey: options.key) != nil
            promise.resolve(containsKey)
        }.runOnQueue(.main)
    }
}
